local_planner/src/local_footstep_planner.cpp

```diff
@@ -589,3 +577,92 @@ Eigen::Vector3d LocalFootstepPlanner::getNearestValidFoothold(
+                    R.at(1).y() * R.at(2).y() * R.at(2).y())) /
+             (R.at(0).x() * R.at(1).y() - R.at(1).x() * R.at(0).y() -
+              R.at(0).x() * R.at(2).y() + R.at(2).x() * R.at(0).y() +
+              R.at(1).x() * R.at(2).y() - R.at(2).x() * R.at(1).y()),
+        (0.5000 * (-R.at(0).x() * R.at(0).x() * R.at(1).x() +
+                   R.at(0).x() * R.at(0).x() * R.at(2).x() +
+                   R.at(0).x() * R.at(1).x() * R.at(1).x() -
+                   R.at(0).x() * R.at(2).x() * R.at(2).x() +
+                   R.at(0).x() * R.at(1).y() * R.at(1).y() -
+                   R.at(0).x() * R.at(2).y() * R.at(2).y() -
+                   R.at(1).x() * R.at(1).x() * R.at(2).x() +
+                   R.at(1).x() * R.at(2).x() * R.at(2).x() -
+                   R.at(1).x() * R.at(0).y() * R.at(0).y() +
+                   R.at(1).x() * R.at(2).y() * R.at(2).y() +
+                   R.at(2).x() * R.at(0).y() * R.at(0).y() -
+                   R.at(2).x() * R.at(1).y() * R.at(1).y())) /
+            (R.at(0).x() * R.at(1).y() - R.at(1).x() * R.at(0).y() -
+             R.at(0).x() * R.at(2).y() + R.at(2).x() * R.at(0).y() +
+             R.at(1).x() * R.at(2).y() - R.at(2).x() * R.at(1).y()),
+        (0.5000 *
+         std::sqrt((R.at(0).x() * R.at(0).x() - 2 * R.at(0).x() * R.at(1).x() +
+                    R.at(1).x() * R.at(1).x() + R.at(0).y() * R.at(0).y() -
+                    2 * R.at(0).y() * R.at(1).y() + R.at(1).y() * R.at(1).y()) *
+                   (R.at(0).x() * R.at(0).x() - 2 * R.at(0).x() * R.at(2).x() +
+                    R.at(2).x() * R.at(2).x() + R.at(0).y() * R.at(0).y() -
+                    2 * R.at(0).y() * R.at(2).y() + R.at(2).y() * R.at(2).y()) *
+                   (R.at(1).x() * R.at(1).x() - 2 * R.at(1).x() * R.at(2).x() +
+                    R.at(2).x() * R.at(2).x() + R.at(1).y() * R.at(1).y() -
+                    2 * R.at(1).y() * R.at(2).y() +
+                    R.at(2).y() * R.at(2).y()))) /
+            std::abs(R.at(0).x() * R.at(1).y() - R.at(1).x() * R.at(0).y() -
+                     R.at(0).x() * R.at(2).y() + R.at(2).x() * R.at(0).y() +
+                     R.at(1).x() * R.at(2).y() - R.at(2).x() * R.at(1).y());
+
+    return D;
+  }
+
+  if (P.empty()) {
+    // Trivial
+    Eigen::Vector3d D;
+    switch (R.size()) {
+      case 0:
+        D << 0, 0, 0;
+        break;
+      case 1:
+        D << R.at(0).x(), R.at(0).y(), 0;
+        break;
+      case 2:
+        D << (R.at(0).x() + R.at(1).x()) / 2, (R.at(0).y() + R.at(1).y()) / 2,
+            (R.at(0) - R.at(1)).norm() / 2;
+        break;
+
+      default:
+        D << 0, 0, 0;
+        break;
+    }
+
+    return D;
+  }
+
+  // Recursive
+  Eigen::Vector2d p = P.back();
+  P.pop_back();
+  Eigen::Vector3d D = welzlMinimumCircle(P, R);
+  if ((p - D.segment(0, 2)).norm() < D.z()) {
+    return D;
+  }
+
+  R.push_back(p);
+  return welzlMinimumCircle(P, R);
+}
```

Comment:
Have we benchmarked how long computeFootPlan takes with these new additions?

Reply:
I can benchmark it. But this algorithm is recursive and the time complexity is O(N). Since we entered a stance plan of about 6 elements, it basically calculates the circumcircle of 3 points and computes the distance of the remaining points.

Reply:
The circumcircle looks complicated because I copied it directly from MATLAB. I suspect it might be more simplified if I generate matlabfunction.

Reply:
It takes between 0.05 and 0.1 ms on my laptop.